### ISSUE TITLE
32-bit build failure: no matching member function for call to 'lerp_table_lookup_float'

### ISSUE THREAD

**Issue by ryandesign** [LMMS/lmms#2044]
Hello, I'm the maintainer of lmms for the MacPorts package management system. I noticed that on my 64-bit Mac, I can build lmms 1.1.3 64-bit but not 32-bit. Here are the warnings and errors that were displayed:

```
[ 77%] Building CXX object plugins/LadspaEffect/calf/CMakeFiles/calf.dir/src/organ.o
cd /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/build-i386/plugins/LadspaEffect/calf && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++   -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_XML_LIB -Dcalf_EXPORTS -O2 -g -fno-exceptions -Wall -Werror=unused-function -Wno-sign-compare -Wno-strict-overflow -pipe -Os -arch i386 -stdlib=libc++ -isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk  -fPIC -DPIC -DNDEBUG -arch i386 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk -mmacosx-version-min=10.10 -fPIC -isystem /opt/local/include -isystem /opt/local/include/QtGui -isystem /opt/local/include/QtXml -isystem /opt/local/include/QtCore -I/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/include -I/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/build-i386 -I/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf -I/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src    -D'QT_TRANSLATIONS_DIR="/opt/local/share/qt4/translations"' -O2 -finline-functions  -o CMakeFiles/calf.dir/src/organ.o -c /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp
clang: warning: optimization flag '-finline-functions' is not supported
clang: warning: argument unused during compilation: '-finline-functions'
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:23:
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/giface.h:25:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/primitives.h:377:5: warning: control may reach end of non-void function [-Wreturn-type]
    }
    ^
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:27:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/metadata.h:54:10: warning: 'calf_plugins::filter_metadata::is_cv' hides overloaded virtual function [-Woverloaded-virtual]
    bool is_cv(int param_no) { return param_no != par_mode && param_no != par_inertia; }
         ^
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/giface.h:608:10: note: hidden overloaded virtual function 'calf_plugins::plugin_metadata<calf_plugins::filter_metadata>::is_cv' declared here: different qualifiers (const vs none)
    bool is_cv(int param_no) const { return true; }
         ^
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:27:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/metadata.h:64:10: warning: 'calf_plugins::filterclavier_metadata::is_cv' hides overloaded virtual function [-Woverloaded-virtual]
    bool is_cv(int param_no) { return param_no != par_mode && param_no != par_inertia; }
         ^
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/giface.h:608:10: note: hidden overloaded virtual function 'calf_plugins::plugin_metadata<calf_plugins::filterclavier_metadata>::is_cv' declared here: different qualifiers (const vs none)
    bool is_cv(int param_no) const { return true; }
         ^
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:27:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/metadata.h:502:26: warning: 'calf_plugins::organ_metadata::get_commands' hides overloaded virtual function [-Woverloaded-virtual]
    plugin_command_info *get_commands();
                         ^
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/giface.h:605:26: note: hidden overloaded virtual function 'calf_plugins::plugin_metadata<calf_plugins::organ_metadata>::get_commands' declared here: different qualifiers (const vs none)
    plugin_command_info *get_commands() const { return NULL; }
                         ^
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:156:19: error: no matching member function for call to 'lerp_table_lookup_float'
        return ph.lerp_table_lookup_float(data);
               ~~~^~~~~~~~~~~~~~~~~~~~~~~
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/fixed_point.h:233:14: note: candidate template ignored: substitution failure [with U = float]: array is too large (2147483649 elements)
    inline U lerp_table_lookup_float(U data[(unsigned int)(1<<IntBits)+1]) const {
             ^                             ~
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:160:19: error: no matching member function for call to 'lerp_table_lookup_float_mask'
        return ph.lerp_table_lookup_float_mask(data, ORGAN_BIG_WAVE_SIZE - 1);
               ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/fixed_point.h:239:14: note: candidate template ignored: substitution failure [with U = float]: array is too large (2147483649 elements)
    inline U lerp_table_lookup_float_mask(U data[(unsigned int)(1<<IntBits)+1], unsigned int mask) const {
             ^                                  ~
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:341:10: warning: 'calf_plugins::organ_audio_module::is_cv' hides overloaded virtual function [-Woverloaded-virtual]
    bool is_cv(int param_no) { return false; }
         ^
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/giface.h:608:10: note: hidden overloaded virtual function 'calf_plugins::plugin_metadata<calf_plugins::organ_metadata>::is_cv' declared here: different qualifiers (const vs none)
    bool is_cv(int param_no) const { return true; }
         ^
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:343:10: warning: 'calf_plugins::organ_audio_module::is_noisy' hides overloaded virtual function [-Woverloaded-virtual]
    bool is_noisy(int param_no) { return true; }
         ^
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/giface.h:609:10: note: hidden overloaded virtual function 'calf_plugins::plugin_metadata<calf_plugins::organ_metadata>::is_noisy' declared here: different qualifiers (const vs none)
    bool is_noisy(int param_no) const { return false; }
         ^
In file included from /opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/organ.cpp:24:
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:354:18: warning: 'calf_plugins::organ_audio_module::pitch_bend' hides overloaded virtual function [-Woverloaded-virtual]
    virtual void pitch_bend(int /*channel*/, int value) { dsp::drawbar_organ::pitch_bend(value); }
                 ^
/opt/local/var/macports/build/_Users_rschmidt_macports_dports_multimedia_lmms/lmms/work/lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/organ.h:307:10: note: hidden overloaded virtual function 'dsp::drawbar_organ::pitch_bend' declared here: different number of parameters (1 vs 2)
    void pitch_bend(int amt);
         ^
7 warnings and 2 errors generated.
```

Some possibly relevant system info:

```
$ clang -v -arch i386
Apple LLVM version 6.1.0 (clang-602.0.49) (based on LLVM 3.6.0svn)
Target: i386-apple-darwin14.4.0
Thread model: posix
$
```
```
$ xcodebuild -version
Xcode 6.3.1
Build version 6D1002
$
```
```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.10.4
BuildVersion:	14E11f
$
```

**Comment by tresf** [comment]
@ryandesign thanks for the bug report.  Since you've already issued a PR over at #2045, I assume you're on top of this.  In regards to the 32-bit detection, GCC seems to have a "wontfix" bug report about that behavior, so the fix you supplied may serve to benefit Clang compilers only, but if there's no harm merging, we can get it upstream.

**Offtopic**

1. We've switched to building our code on Lion which seems to fix some miscellaneous issues and offer much wider platform compatibility over the previous Mavericks binaries, but encounter this: https://github.com/LMMS/lmms/issues/1633#issuecomment-70040535.

2. Thanks for help keeping the MacPorts build alive.  Many don't know it, but it was MacPorts that made most of the bugs with our binary release possible. :+1:

**Comment by ryandesign** [comment]
#2045 doesn't fix this problem, and I don't know what does; I'm not really a C programmer.

**Comment by tresf** [comment]
This code is actually part of the Calf project and AFAIK 32-bit Apple (Snow Leopard) isn't something we ever plan to support, so the quickest approach right now is to diff the offending plugins to "OFF" via something like:

https://github.com/LMMS/lmms/blob/master/CMakeLists.txt#L29

I'm not sure if this will help though as progressive code changes such as in #1633 will just continue to make this effort harder to support.  Opinions welcome of course. :)

**Comment by ryandesign** [comment]
While it's true that Mac OS X 10.6 Snow Leopard is the last version that supports 32-bit Intel Macs, it is still possible and sometimes useful to build software in 32-bit mode on newer versions of OS X, including the current version 10.10. Since your cmake files attempt to check if the system is 32-bit or 64-bit, I assumed you intended your code to be buildable on 32-bit and 64-bit systems. Are you able to build this code for 32-bit on other operating systems?

By the way, lmms 1.1.3 in MacPorts built fine on all of our automated builders, including the 10.6 x86_64 builder, though I have no tried to actually run lmms on anything earlier than 10.10.

**Comment by tresf** [comment]
> Are you able to build this code for 32-bit on other operating systems?

Yes, 32-bit Windows and Linux build fine.  These LADSPA Calf plugins have been around for quite a while, mostly untouched.

> By the way, lmms 1.1.3 in MacPorts built fine on all of our automated builders, including the 10.6 x86_64 builder,

Correct.  The discussion I had linked about initializer lists was in regards to our master (i.e future 1.2) branch.  Unless we di the suggested workaround, 1.2 won't build on Lion any longer.

**Comment by tresf** [comment]
Can you ammend the original bug report with steps to reproduce and also cleanup the non-relevant warnings from the log?

I tried reproducing this tonight to no avail.  I assume there are some CMAKE flags to set, but the ones I tried from #2045 couldn't get past 0%.

-Tres

**Comment by softrabbit** [comment]
```
lmms-1.1.3/plugins/LadspaEffect/calf/src/calf/fixed_point.h:233:14: note: candidate template ignored: substitution failure [with U = float]: array is too large (2147483649 elements)
    inline U lerp_table_lookup_float(U data[(unsigned int)(1<<IntBits)+1]) const {
```
That is one huge lookup table, 8 GB if a float is 32 bits... :open_mouth: 

The plugin that the _errors_ here relate to is the CALF organ plugin that LMMS can't use anyway, so an easy fix could be to disable it for the time being.

**Comment by ryandesign** [comment]
Why does lmms build things it doesn't need? :smile: 

I have now set up a 32-bit Intel iMac with OS X v10.6.8, and was able to compile lmms 1.1.3 from MacPorts just fine using the default compiler on that system (Apple gcc-4.2.1) and it ran and played the demo songs fine too. So there doesn't seem to be anything inherently 32-bit-incompatible here, I suspect there's just something about this code that clang interprets differently from gcc. Maybe the code is inadvertently using gcc-specific extensions of some sort. I do see that these are declared `inline`, and I do know that clang does handle `inline` differently from gcc. There is some [documentation on that](http://clang.llvm.org/compatibility.html#inline), but that mainly talks about the difference between clang's default C99 mode and gcc's default gnu89 mode, which isn't applicable here because in MacPorts we're already instructing lmms to build in gnu89 mode to fix another problem. It also says this applies only to C code, and it looks like the problem code here is C++, in which it says `inline` is handled differently again.

**Comment by tresf** [comment]
I've tackled some of the `inline` bugs in the swh plugins through a very painstaking trial and error via stuff like this: https://github.com/swh/ladspa/commit/d07c1cbb2fd60a0a552982d58e0692994a138469.

But that was just one snag... I also hit a for-loop problem per: https://github.com/swh/ladspa/commit/79864921afa98bfd7430028cd4606040a81c1532

And some mystery problems like this if/else block:
https://github.com/swh/ladspa/commit/bb073f6fcf0c464baf86a3606992858b3c0645a6

But again, without the necessary steps to compile (https://github.com/LMMS/lmms/issues/2044#issuecomment-101120361), I can't help with this one.

> Why does lmms build things it doesn't need? :smile:

Someone with a better knowledge of our plugins and how to safely strip out `organ.cpp` and `organ.h` could easily do this, but is this the only compilation issue?

I'm impressed by the dedication to this issue, but I still have to question who this fix benefits, as I haven't seen a 32-bit Mac in quite a few years.  Does MacPorts still have a lot of 32-bit users?

**Comment by Wallacoloo** [comment]
Hmm, so the first actual error:
```
organ.h:156:19: error: no matching member function for call to 'lerp_table_lookup_float'
        return ph.lerp_table_lookup_float(data);
```

corresponds to [this](https://github.com/LMMS/lmms/blob/master/plugins/LadspaEffect/calf/src/calf/organ.h#L156) code:

```c++
    inline float wave(float *data, dsp::fixed_point<int, 20> ph) {
        return ph.lerp_table_lookup_float(data);
    }
```

And the fixed_point class starts off with:
```c++
template<class T, int FracBits> class fixed_point {
    T value;
    enum { IntBits = (sizeof(T)/8) - FracBits };
```

Also, the function definition that it fails to match because of a ridiculous array bounds of 2147483649 (that's `0b10000000000000000000000000000001` in binary) is ([fixed_point.h:233](https://github.com/LMMS/lmms/blob/master/plugins/LadspaEffect/calf/src/calf/fixed_point.h#L233)):
```c++
template<class U> 
    inline U lerp_table_lookup_float(U data[(unsigned int)(1<<IntBits)+1]) const {
...
```
Note that's a member function of the fixed_point template class.

So in this case, IntBits should be `(4/8) - 20` = `-20`, assuming `sizeof(int)` == 4 bytes. I would think that the author intended for IntBits to be set to `(sizeof(T)*8) - FracBits` instead. Although I wouldn't have expected `(unsigned int)(1<<-20)+1` to result in the value `2147483649`, [bit shifting by a negative shift amount is undefined behavior in C](http://stackoverflow.com/a/4945765/216292). So that could very well be the error. This is quite possibly something that different gcc versions happen to treat differently.

**Comment by Wallacoloo** [comment]
So I whipped up a quick test. If you compile [this](https://gist.github.com/Wallacoloo/ea79a5ed52d0e80ccc14) file, which makes use of the `fixed_point` in the same way as the Calf code, `(unsigned int)(1<<IntBits)+1` comes out to `1` when using gcc 4.9 or 4.8 on my Linux system, and clang++-3.6, 3.5 and 3.4 always print a different value for `(unsigned int)(1<<IntBits)+1` (perhaps recognizing that it's undefined behavior and thus applying some clever optimization). I cannot reproduce the compile-time error, but the value of IntBits definitely seems to be incorrect.

**Comment by Wallacoloo** [comment]
Dur, compiling the linked example with the `-m32` flag gives this (ignore that the file is named test.cpp instead of main.cpp):
```
clang++ -m32 ./test.cpp 
./test.cpp:281:15: error: no matching member function for call to 'lerp_table_lookup_float'
    return ph.lerp_table_lookup_float(data);
           ~~~^~~~~~~~~~~~~~~~~~~~~~~
./test.cpp:242:14: note: candidate template ignored: substitution failure [with U = float]: array is too large (2147483649 elements)
    inline U lerp_table_lookup_float(U data[(unsigned int)(1<<IntBits)+1]) const {
             ^                             ~
./test.cpp:52:31: warning: format specifies type 'long' but the argument has underlying type 'unsigned int' [-Wformat]
        printf("IntBits: %ld\n", IntBits);
                         ~~~     ^~~~~~~
                         %u
./test.cpp:285:31: note: in instantiation of member function 'dsp::fixed_point<int, 20>::fixed_point' requested here
        float f = wave((float*)NULL, dsp::fixed_point<int, 20>());
                                     ^
./test.cpp:53:93: warning: shift count is negative [-Wshift-count-negative]
        printf("size of array with bounds (unsigned int)(1<<IntBits)+1: %u\n", (unsigned int)(1<<IntBits)+1);
                                                                                               ^ ~~~~~~~
2 warnings and 1 error generated.
```

Notice that the array size is the same as originally reported in this thread, and I am using Ubuntu. So this does not seem to be a mac-specific error. g++ will still happily compile this with only a warning.

**Comment by ryandesign** [comment]
> But again, without the necessary steps to compile (#2044 (comment)), I can't help with this one.

A reproduction recipe using MacPorts is, on a 64-bit Mac with OS X 10.7 or later, to install all lmms dependencies universal (i.e. for 32-bit and 64-bit), which I think the following command should be sufficient to do:

```
sudo port selfupdate
sudo port install rdepof:lmms +universal
```

Then, build lmms 32-bit:

```
sudo port -d build lmms configure.build_arch=i386
```

> Someone with a better knowledge of our plugins and how to safely strip out organ.cpp and organ.h could easily do this, but is this the only compilation issue?

Don't know, but I'll make the change suggested by Wallacoloo and see if anything else stops the build.

> I'm impressed by the dedication to this issue, but I still have to question who this fix benefits, as I haven't seen a 32-bit Mac in quite a few years. Does MacPorts still have a lot of 32-bit users?

This fix benefits users with 32-bit Macs, because it hopefully eventually lets you compile a 32-bit/64-bit universal lmms application that can be used on 32-bit or 64-bit Macs. The first batch of Intel Macs produced in 2006 used 32-bit processors ([Core Duo](http://www.everymac.com/systems/by_processor/intel-core-duo-macs.html) and [Core Solo](http://www.everymac.com/systems/by_processor/intel-core-solo-macs.html)). They may be old, but I'm sure many are not broken, and people may want to repurpose old machines as secondary machines performing new and exciting purposes. Perhaps someone wants to repurpose their old iMac as a music machine.

We still get bug reports from users running OS X 10.4 and 10.5 on PowerPC machines. If the hardware works, why should they throw it away when they can still install some useful software on it?

Also, it's not even solely about getting lmms to work on older machines. It's also about fixing problems. In this case, it looks like @Wallacoloo found code that's just wrong, on any platform, so now it can be fixed! Hooray!

> ```
>     enum { IntBits = (sizeof(T)/8) - FracBits };
> ```

> I would think that the author intended for IntBits to be set to `(sizeof(T)*8) - FracBits` instead.

Good catch! I wondered if `IntBits` might be getting computed wrong, but I didn't look at the definition long enough to see what was wrong with it. I'll try this change.

**Comment by softrabbit** [comment]
Maybe time to sync changes from upstream? That left shift bug is fixed there: http://sourceforge.net/p/calf/code/ci/aaca6447de6ed07bb39f376095bb246cc8f3580a/tree//src/calf/fixed_point.h?diff=d727afe27999f888452204cc09fb3b190741ab53

**Comment by ryandesign** [comment]
They've switched to github: https://github.com/calf-studio-gear/calf

The above diff is now available at: https://github.com/calf-studio-gear/calf/commit/aaca6447de6ed07bb39f376095bb246cc8f3580a

So, what's involved in updating lmms' copy of calf to the latest upstream code?

**Comment by softrabbit** [comment]
> So, what's involved in updating lmms' copy of calf to the latest upstream code?

At a glance, it looks like LMMS is using just the ```src``` directory of the calf tree, with ```CMakeLists.txt``` instead of the ```Autotools``` stuff, could be a matter of copying files. 

There might be some LMMS-specific changes to sort out, though.

**Comment by Wallacoloo** [comment]
@tresf just applied the upstream fix for the IntBits issue: https://github.com/LMMS/lmms/commit/5375bda6ac9a155173f007f4c8cd0276270acca1

So hopefully this particular issue is now fixed in master.

**Comment by tresf** [comment]
Thanks, forgot to revisit this.  Closed via 5375bda.  Please reopen if you believe this was closed in error.
